fastly/custom_tls_activation.go

```diff
@@ -0,0 +1,195 @@
+	PageNumber               *uint   // The page index for pagination.
+	PageSize                 *uint   // The number of activations per page.
+}
+
+// formatFilters converts user input into query parameters for filtering.
+func (i *ListTLSActivationsInput) formatFilters() map[string]string {
+	result := map[string]string{}
+	pairings := map[string]interface{}{
+		"filter[tls_certificate.id]":   i.FilterTLSCertificateID,
+		"filter[tls_configuration.id]": i.FilterTLSConfigurationID,
+		"filter[tls_domain.id]":        i.FilterTLSDomainID,
+		"include":                      i.Include,
+		"page[number]":                 i.PageNumber,
+		"page[size]":                   i.PageSize,
+	}
+	for key, value := range pairings {
+		if !reflect.ValueOf(value).IsNil() {
+			result[key] = fmt.Sprintf("%v", reflect.ValueOf(value).Elem())
+		}
+	}
+
+	return result
+}
+
+// ListTLSActivations list all activations.
+func (c *Client) ListTLSActivations(i *ListTLSActivationsInput) ([]*TLSActivation, error) {
+
+	p := "/tls/activations"
+	filters := &RequestOptions{
+		Params: i.formatFilters(),
+		Headers: map[string]string{
+			"Accept": "application/vnd.api+json", // this is required otherwise the filters don't work
+		},
+	}
+
+	r, err := c.Get(p, filters)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := jsonapi.UnmarshalManyPayload(r.Body, reflect.TypeOf(new(TLSActivation)))
+	if err != nil {
+		return nil, err
+	}
+
+	a := make([]*TLSActivation, len(data))
+	for i := range data {
+		typed, ok := data[i].(*TLSActivation)
+		if !ok {
+			return nil, fmt.Errorf("got back a non-TLSActivation response")
+		}
+		a[i] = typed
+	}
+
+	return a, nil
+}
+
+// GetTLSActivationInput is used as input to the GetTLSActivation function.
+type GetTLSActivationInput struct {
+	ID string
+}
+
+// GetTLSActivation retrieve a single activation.
+func (c *Client) GetTLSActivation(i *GetTLSActivationInput) (*TLSActivation, error) {
+
+	if i.ID == "" {
+		return nil, ErrMissingID
+	}
+
+	p := fmt.Sprintf("/tls/activations/%s", i.ID)
+
+	r, err := c.Get(p, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var a TLSActivation
+	if err := jsonapi.UnmarshalPayload(r.Body, &a); err != nil {
+		return nil, err
+	}
+
+	return &a, nil
+}
+
+// CreateTLSActivationInput is used as input to the CreateTLSActivation function.
+type CreateTLSActivationInput struct {
+	TLSCertificate   *TLSCertificate   `jsonapi:"relation,tls_certificate,tls_certificate"`
+	TLSConfiguration *TLSConfiguration `jsonapi:"relation,tls_configuration,tls_configuration"`
+	TLSDomain        *TLSDomain        `jsonapi:"relation,tls_domain,tls_domain"`
+	Type             string            `jsonapi:"primary,tls_activation"` // Type value does not need to be set but existence of this key prevents server error due to API bug that requires "type" to be present.
```

Comment:
The way [the code](https://github.com/google/jsonapi/blob/9246c912f5ff63f2e315efd188d254f93e5598f0/response.go#L286) is written, it can be named anything as long as it has `primary`. However, using `ID` is preferable here since it has [`omitempty`](https://github.com/google/jsonapi/blob/9246c912f5ff63f2e315efd188d254f93e5598f0/node.go#L39) and the value in this field will be marshaled there.

Also a quick note that the API requiring a `type` field is adherence to the [JSON-API.org specification](https://jsonapi.org/format/#crud-creating), for better or for worse:

> A resource can be created by sending a POST request to a URL that represents a collection of resources. The request MUST include a single resource object as primary data. The resource object MUST contain at least a type member.

Reply:
Agree on the use of `ID` for consistency. I believe in the case of the Update struct these annotations can just be placed on the existing ID field and it will serialize as expected instead of having a `Type` field too

```
// UpdateTLSActivationInput is used as input to the UpdateTLSActivation function.
type UpdateTLSActivationInput struct {
	ID             string          `jsonapi:"primary,tls_activation"`
	TLSCertificate *TLSCertificate `jsonapi:"relation,tls_certificate,tls_certificate"`
}
```

Reply:
Thank you! I'll update per this feedback

Reply:
@thommahoney Noted re: JSON-API spec. I considered it a bug because /tls/configuration and other APIs do not return a server error if it is blank.

Reply:
Understood, and thanks for clarifying here and in the code. 👍